randutil/crypto.go

```diff
@@ -0,0 +1,61 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package randutil
+
+import (
+	cryptorand "crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+)
+
+// CryptoTokenBytes returns a crypthographically secure token of
+// nbytes random bytes.
+func CryptoTokenBytes(nbytes int) ([]byte, error) {
+	b := make([]byte, nbytes)
+	_, err := cryptorand.Read(b)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain %d crypto random bytes: %v", nbytes, err)
+	}
+	return b, nil
+}
+
+// CryptoToken returns a crypthographically secure token string
+// encoding nbytes random bytes.
+// The result is URL-safe.
+func CryptoToken(nbytes int) (string, error) {
+	b, err := CryptoTokenBytes(nbytes)
+	if err != nil {
+		return "", err
+	}
+	return base64.RawURLEncoding.EncodeToString(b), nil
+}
+
+// RandomKernelUUID will return a UUID from the kernel's procfs API at
+// /proc/sys/kernel/random/uuid. Only to be used in very specific uses, most
+// random code should use CryptoToken(Bytes) instead.
+func RandomKernelUUID() string {
```

Comment:
Slightly sad that this is not tested, but I guess it's not entirely trivial as not every test environment will have access to /proc/sys/kernel/random/uuid but there could be a c.Skip() if that is not available. We should probably add a test in snapd and cherry pick here but that should not block.

Reply:
I will look into this, as a separate PR!